pytorch_lightning/trainer/lr_finder.py

```diff
@@ -0,0 +1,288 @@
+        self.accumulate_grad_batches = num_accumulation_steps
+
+        # Configure optimizer and scheduler
+        optimizer, _, _ = self.init_optimizers(model.configure_optimizers())
+        assert len(optimizer) == 1, 'cannot find lr for more than 1 optimizer'
+        configure_optimizers = model.configure_optimizers
+        model.configure_optimizers = lr_finder._get_new_optimizer(optimizer[0])
+
+        # Fit, lr & loss logged in callback
+        self.fit(model)
+
+        # Promt if we stopped early
+        if self.global_step != num_iters:
+            print('LR finder stopped early due to diverging loss.')
+
+        # Transfer results from callback to lr finder object
+        lr_finder.results.update({'lr': self.callbacks[0].lrs,
+                                  'loss': self.callbacks[0].losses})
+
+        # Finish by resetting variables so trainer is ready to fit model
+        self.logger = logger
+        self.callbacks = callbacks
+        self.max_steps = max_steps
+        self.show_progress_bar = show_progress_bar
+        self.accumulate_grad_batches = accumulate_grad_batches
+        model.configure_optimizers = configure_optimizers
+
+        return lr_finder
+
+
+class _LRFinder(object):
+    def __init__(self, mode, lr_min, lr_max, num_iters):
+        assert mode in ('linear', 'exponential'), \
+            'mode should be either `linear` or `exponential`'
+
+        self.mode = mode
+        self.lr_min = lr_min
+        self.lr_max = lr_max
+        self.num_iters = num_iters
+
+        self.results = {}
+
+    def _get_new_optimizer(self, optimizer):
+        new_lrs = [self.lr_min] * len(optimizer.param_groups)
+        for param_group, new_lr in zip(optimizer.param_groups, new_lrs):
+            param_group["lr"] = new_lr
+            param_group["initial_lr"] = new_lr
+
+        args = (optimizer, self.lr_max, self.num_iters)
+        scheduler = _LinearLR(*args) if self.mode == 'linear' else _ExponentialLR(*args)
+
+        def configure_optimizers():
+            return [optimizer], [{'scheduler': scheduler,
+                                  'interval': 'step'}]
+
+        return configure_optimizers
+
+    def plot(self, suggest=False, ax=None):
+        import matplotlib.pyplot as plt
+
+        lrs = self.results["lr"]
+        losses = self.results["loss"]
+
+        fig = None
+        if ax is None:
+            fig, ax = plt.subplots()
+
+        # Plot loss as a function of the learning rate
+        ax.plot(lrs, losses)
+        if self.mode == 'exponential':
+            ax.set_xscale("log")
+        ax.set_xlabel("Learning rate")
+        ax.set_ylabel("Loss")
+
+        if suggest:
+            _ = self.suggestion()
+            if self._optimal_idx:
+                ax.plot(lrs[self._optimal_idx], losses[self._optimal_idx],
+                        markersize=10, marker='o', color='red')
+
+        if fig is not None:
+            plt.show()
```

Comment:
I would rather just return the figure object which also allows exporting it

Reply:
I have added a `show` argument which by default is `False`, still giving people the option to show it.
Is it fine to return the `ax` handle or should `fig` also be returned?

Reply:
most people are running this in a (remote) terminal and the show may crash beast of missing libs or silently open to and wait till user close it but user does not see anything...
so I would prefer just return the figure object... `ax = fig.gca()` allows you to edit it further later